### PR TITLE
Generate observations using mapping and csvData

### PR DESCRIPTION
**Pull request by pradh**
The main API returns observations as a map, associated with rows, so that the frontend can decide to link it to display rows.

There's also a helper to produce a text string from the map that's to be tweaked as necessary.

### CODE REVIEW

File: wizard/src/utils/obs_generation.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      }
+    } else if (mval.type == MappingType.CONSTANT) {
+      obsGenMaps.thing2Const.set(mthing, mval.constant);
+    }
+  }
+
+  const rowObs: RowObservations = new Map();
+  for (const idx of Array.from(csvData.rowsForDisplay.keys())) {
+    const row = csvData.rowsForDisplay.get(idx);
+    const obs = generateObservationsInRow(
```
Comment (chejennifer) [inline wizard/src/utils/obs_generation.ts:73]: nit: functions should be defined before they are used, so `generateObservationsInRow` should be defined before this current function

File: wizard/src/utils/obs_generation.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  row: Array<string>,
+  header: Array<Column>,
+  obsGenMaps: ObsGenMaps
+): Array<Observation> {
+  const obsList = new Array<Observation>();
+  for (const valColIdx of Array.from(obsGenMaps.valCol2Hdr.keys())) {
+    const hdrThing = obsGenMaps.valCol2Hdr.get(valColIdx);
+    const obs: Observation = new Map();
+    // TODO: Consider if we want to do some cleaning and checking for non-numeric value.
+    if (row[valColIdx] != null && row[valColIdx] !== "") {
```
Comment (chejennifer) [inline wizard/src/utils/obs_generation.ts:111]: can actually just use the lodash function `_.isEmpty()`: https://lodash.com/docs/4.17.15#isEmpty

File: wizard/src/utils/obs_generation.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  Mapping,
+  MappingType,
+  MappingVal,
+  Observation,
+  RowObservations,
+} from "../types";
+
+// Helper maps used while generating observaitons.
+interface ObsGenMaps {
+  // Column Index -> mapping thing found in header for VALUE.
```
Comment (chejennifer) [inline wizard/src/utils/obs_generation.ts:30]: nit: "mapped thing found .."

**Review by chejennifer — COMMENTED** [review]
this is great!

File: wizard/src/utils/obs_generation.ts (hunk 3 above)
Comment (pradh) [reply wizard/src/utils/obs_generation.ts]: Done.

File: wizard/src/utils/obs_generation.ts (hunk 1 above)
Comment (pradh) [reply wizard/src/utils/obs_generation.ts:122]: I see. Moved it around.

File: wizard/src/utils/obs_generation.ts (hunk 2 above)
Comment (pradh) [reply wizard/src/utils/obs_generation.ts]: Nice tip!

**Review by pradh — COMMENTED** [review]
Thanks for the review Jennifer!

File: wizard/src/utils/obs_generation.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  CsvData,
+  MappedThing,
+  Mapping,
+  MappingType,
+  MappingVal,
+  Observation,
+  RowObservations,
+} from "../types";
+
+// Helper maps used while generating observaitons.
```
Comment (jehangiramjad) [inline wizard/src/utils/obs_generation.ts:30]: nit: typo at the end of this comment. s/observaitons/observations

File: wizard/src/utils/obs_generation.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    if (hasRequiredProps(obs)) {
+      obsList.push(obs);
+    }
+  }
+  return obsList;
+}
+
+/**
+ * Given a mapping and CSV Data computes the observations for every row
+ * corresponding to |rowsForDisplay|.
```
Comment (jehangiramjad) [inline wizard/src/utils/obs_generation.ts:89]: nit: this function comment does not seem to follow the function comment format with @param and @returns etc

File: wizard/src/utils/obs_generation.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    );
+    if (obs.length > 0) {
+      rowObs.set(idx, obs);
+    }
+  }
+  return rowObs;
+}
+
+/**
+ * Given an Observation that was returned from generateRowObservations(), returns
```
Comment (jehangiramjad) [inline wizard/src/utils/obs_generation.ts:135]: ditto the nit comment about comment format above.

**Review by jehangiramjad — APPROVED** [review]
Looks good! Just a couple of nitty comments.

File: wizard/src/utils/obs_generation.ts (hunk 4 above)
Comment (pradh) [reply wizard/src/utils/obs_generation.ts]: fixed.

File: wizard/src/utils/obs_generation.ts (hunk 5 above)
Comment (pradh) [reply wizard/src/utils/obs_generation.ts:89]: Thanks for the reminder :)

File: wizard/src/utils/obs_generation.ts (hunk 6 above)
Comment (pradh) [reply wizard/src/utils/obs_generation.ts:139]: Done.

**Review by pradh — COMMENTED** [review]
Thank you for the comments Jehangir!

**Comment by pradh** [comment]
@chejennifer: I have addressed your comments.  Going to submit this to unblock things next week.

Please add any further comments you have, and I will address them in a follow up PR...

**Pull request merged**